test/distributed/test_distributed.py

```diff
@@ -2750,33 +2755,84 @@ def test_ddp_grad_div_uneven_inputs(self):
         )
         n_iters = 1 if self.rank == 0 else 2
-        from torch.nn.parallel.distributed import join
 
-        with join(net) as ddp_join:
+        with net.join(net):
             for i in range(n_iters):
                 loss = net(inp).sum()
                 loss.backward()
                 # The grad is always expected_grad, since we divide by the number
                 # of currently active processes and inactive processes contribute
                 # zero gradient. If we kept dividing by static initial world
                 # size as processes leave, the grad would be smaller.
-                expected_grad = torch.ones(dim, dim).to(self.rank) * grad_scale
+                expected_grad = torch.ones(dim, dim, device=self.rank) * grad_scale
                 param = list(net.parameters())[0]
                 self.assertEqual(expected_grad, param.grad)
                 # Avoid accumulating grads so that it's the same every iteration
                 net.zero_grad()
                 torch.cuda.synchronize(device=self.rank)
 
+    @require_backend({"gloo", "nccl"})
+    @require_backends_available({"gloo", "nccl"})
+    @skip_if_lt_x_gpu(2)
+    @skip_if_rocm
+    def test_ddp_join_model_equivalence(self):
+        # Verifies equivalence with model training locally and with DDP under
+        # the join context manager.
+        batch = 3
+        dim = 10
+        learning_rate = 0.03
+        model = nn.Linear(dim, dim, bias=False)
+        inp = torch.randn(batch, dim, device=self.rank)
+        local_model = copy.deepcopy(model)
+        local_model = local_model.cuda(self.rank)
+        rank_to_iter_mapping = {rank : 2 * rank for rank in range(dist.get_world_size())}
+        # run local model
+        local_iters = max(rank_to_iter_mapping.values())
+
+        for _ in range(local_iters):
+            out = local_model(inp)
+            loss = out.sum()
+            loss.backward()
+            for param in local_model.parameters():
+                if param.grad is not None:
+                    with torch.no_grad():
+                        param.add_(param.grad, alpha=learning_rate)
+                        param.grad.detach_()
+                        param.grad.zero_()
+
+        # run DDP model with join API
+        num_iters = rank_to_iter_mapping[self.rank]
+        net = torch.nn.parallel.DistributedDataParallel(
+            model.cuda(self.rank), device_ids=[self.rank]
+        )
+
+        with net.join():
+            for _ in range(num_iters):
+                out = net(inp)
+                loss = out.sum()
+                loss.backward()
+                torch.cuda.synchronize()
+                for param in net.module.parameters():
+                    if param.grad is not None:
+                        with torch.no_grad():
+                            param.add_(param.grad, alpha=learning_rate)
+                            param.grad.detach_()
+                            param.grad.zero_()
```

Comment:
nit: any reason we're not just using a simple optimizer here instead of manually updating the params?

Reply:
I was mostly following the pattern in the other tests in the file, where they manually update grads when running equivalence checks. Although, a vanilla SGD should probably be fine here.